### DISCUSSION TITLE
Ideas from 4/1/2021 Meeting with Bill Schlotter

### DISCUSSION THREAD

**Discussion by ZLLentz** [pcdshub/atef#3] [Ideas]
https://confluence.slac.stanford.edu/display/XBD/Notes+on+Nominal+PVs%2C+Streamlining+checklists+and+ATEF

- ATEF should include a way to restore nominal values of configuration PVs
- ATEF needs some auxiliary PVs for handling configuration controls, timestamps of last restore, timestamp of last change of the nominal
- changing the nominal values must require at least a caput
- config file for adding/removing PVs: it should be easy for a scientist to add additional PVs to be tracked by ATEF
- generate html reports from atef tests, publish them to pswww
- post alerts to the slack channel? Can be done via grafana if there are summary PVs
- Add ATEF summary PVs (per component and also overall)! Is the component ok? Results and timestamps of the last test? Number of passing/failing? Number of values at nominal vs not at nominal?

**Comment by ZLLentz** [comment]
It may be best to implement this as a `caproto` IOC that systematically generates `ATEF` PVs based on user configuration files. This IOC would handle things like:
- Keeping track of nominal value configurations for comparison and restore
- Keeping track of the last time we ran tests and what the results were
- Summarizing test results and nominal value comparisons overall and per-device

So the config file might just be something like a simple yaml file e.g.:
```
im1l0:
    - IM1L0:XTES:CAM:IMAGE1:ArraySize_0
    - ...
```
And then atef-ioc might generate some PVs like
```
IM1L0:XTES:CAM:IMAGE1:ArraySize_0:NOMINAL
ATEF:IM1L0:SUMMARY
```

**Reply by klauer** [reply]
I'm not against using PV names for ATEF in principle, but what if we instead work with happi object names and dotted ophyd attribute names?

I like the IOC PV suggestions overall.

**Reply by ZLLentz** [reply]
I literally just walked back to my computer to type "wait, actually we should use happi and ophyd"
You beat me to it

**Reply by ZLLentz** [reply]
This is especially good because we use the ophyd name in the typhos screen

**Reply by klauer** [reply]
Great, so we both like happi/ophyd for this purpose. 

We should then decide as to whether or not that's the _only_ way you can specify these conditions. That is, do we leave in an "out" for PVs that are not contained in a happi-loaded ophyd Device? I'd like to draw a line here and just say no, but I think there may be some good arguments to leaving it open.

... Or maybe that ends up just looking like ATEF-defined `GenericSignalDevice(pv="pvname", name="pvname")`. Hmm...

**Comment by klauer** [comment]
> changing the nominal values must require at least a caget

`caput`, I'd hope...

> ATEF should include a way to restore nominal values of configuration PVs

This is perhaps the most concerning one to me. I'd assume this only happens at the user's request, but is there any limitation as to _when_ (an exaggerated) `reset_my_hutch.py` can be run?

**Reply by ZLLentz** [reply]
Perhaps we can limit this sort of interaction to very targeted restores e.g.
```
$ atef restore im1l0
Please confirm the following changes:
Field      Old New
Event Code 45  140
```

**Reply by ZLLentz** [reply]
I think it's very clear that ATEF will need detailed logging regardless of other details
